topology/nhlt/nhlt-processor.c

```diff
@@ -0,0 +1,287 @@
+	if (ret < 0)
+		return ret;
+
+	ret = snd_config_add(data, cfg_new);
+	if (ret < 0)
+		snd_config_delete(cfg_new);
+
+	return 0;
+}
+
+static int print_nhlt_as_hex_bytes(struct nhlt *blob, struct endpoint_descriptor **eps, char *dst)
+{
+	uint8_t *top_p = (uint8_t *)blob;
+	struct endpoint_descriptor *ep;
+	uint8_t *ep_p;
+	int i, j;
+
+	for (i = 0; i < sizeof(struct nhlt); i++) {
+		snprintf(dst, 6, "0x%02x,", *top_p);
+		dst += 5;
+		top_p++;
+	}
+
+	for (i = 0; i < blob->endpoint_count; i++) {
+		ep = eps[i];
+		ep_p = (uint8_t *)ep;
+		printf("ep %i length is %u\n", i, ep->length);
+		for (j = 0; j < ep->length; j++) {
+			snprintf(dst, 6, "0x%02x,", *ep_p);
+			dst += 5;
+			ep_p++;
+		}
+	}
+
+	/* remove the last comma... */
+	dst--;
+	*dst = '\0';
+
+	return 0;
+}
+
+/* called at the beginning of topology per-processing */
+int nhlt_init(struct tplg_pre_processor *tplg_pp)
+{
+	if (!tplg_pp)
+		return -EINVAL;
+
+	/* set default values here */
+
+	return 0;
+}
+
+/* called at the beginning of topology per-processing */
+int nhlt_delete(struct tplg_pre_processor *tplg_pp)
+{
+	if (!tplg_pp)
+		return -EINVAL;
+
+	if(tplg_pp->private_data)
+		free(tplg_pp->private_data);
+
+	return 0;
+}
+
+/* called at the end of topology per-processing */
+int nhlt_create(struct tplg_pre_processor *tplg_pp)
+{
+	struct endpoint_descriptor *eps[MAX_ENDPOINT_COUNT];
+	struct intel_nhlt_params *nhlt;
+	char *bytes_string_buffer;
+	int eps_count = 0;
+	struct nhlt blob;
+	size_t nhlt_size;
+	int ret = 0;
+	int i;
+
+	nhlt = (struct intel_nhlt_params *)tplg_pp->private_data;
+
+	if (!nhlt)
+		return -EINVAL;
+
+	for (i = 0; i < MAX_ENDPOINT_COUNT; i++)
+		eps[i] = NULL;
+
+	/* add endpoints here, increase eps_count */
+
+	/* we don't have endpoints */
+	if (!eps_count)
+		return 0;
+
+	blob.efi_acpi.signature = 0;
+	blob.efi_acpi.length = 0;
+	blob.efi_acpi.revision = 0;
+	blob.efi_acpi.checksum = 0;
+	blob.efi_acpi.oem_id[6] = 0;
+	blob.efi_acpi.oem_table_id = 0;
+	blob.efi_acpi.oem_revision = 0;
+	blob.efi_acpi.creator_id = 0;
+	blob.efi_acpi.creator_revision = 0;
+
+	blob.endpoint_count = eps_count;
+
+	/* get blob total size */
+	nhlt_size = sizeof(struct nhlt);
+	for (i = 0; i < eps_count; i++) {
+		if (eps[i])
+			nhlt_size += eps[i]->length;
+	}
+
+	/* add the total length to top level struct */
+	blob.efi_acpi.length = nhlt_size;
+
+	tplg_pp_debug("create_nhlt_blob ep count %d total size %u", eps_count, nhlt_size);
+
+	/* allocate the bytes string, every byte is alsa conf byte string format "0xA1," */
+	bytes_string_buffer = calloc(nhlt_size * 5, sizeof(uint8_t));
+	if (!bytes_string_buffer) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	/* print the whole thing in hex characters */
+	print_nhlt_as_hex_bytes(&blob, eps, bytes_string_buffer);
+
+	add_to_manifest(tplg_pp, bytes_string_buffer);
+
+err:
+	/* remove all enpoints */
+	for (i = 0; i < eps_count; i++) {
+		free(eps[i]);
+		eps[i] = NULL;
+	}
+	eps_count = 0;
```

Comment:
eps_count is a local variable. why do we need to reset this?

Reply:
no need, will remove